upstairs/src/upstairs.rs

```diff
@@ -0,0 +1,3370 @@
+    pub(crate) async fn apply(&mut self, action: UpstairsAction) {
+        match action {
+            UpstairsAction::Downstairs(d) => {
+                self.apply_downstairs_action(d).await
+            }
+            UpstairsAction::Guest(b) => {
+                self.apply_guest_request(b).await;
+                self.gone_too_long();
+            }
+            UpstairsAction::LeakCheck => {
+                const LEAK_MS: usize = 1000;
+                let leak_tick =
+                    tokio::time::Duration::from_millis(LEAK_MS as u64);
+                if let Some(iop_limit) = self.guest.get_iop_limit() {
+                    let tokens = iop_limit / (1000 / LEAK_MS);
+                    self.guest.leak_iop_tokens(tokens);
+                }
+
+                if let Some(bw_limit) = self.guest.get_bw_limit() {
+                    let tokens = bw_limit / (1000 / LEAK_MS);
+                    self.guest.leak_bw_tokens(tokens);
+                }
+
+                self.leak_deadline =
+                    Instant::now().checked_add(leak_tick).unwrap();
+            }
+            UpstairsAction::FlushCheck => {
+                if self.need_flush {
+                    self.submit_flush(None, None).await;
+                }
+                self.flush_deadline = deadline_secs(self.flush_timeout_secs);
+            }
+            UpstairsAction::StatUpdate => {
+                self.on_stat_update().await;
+                self.stat_deadline = deadline_secs(STAT_INTERVAL_SECS);
+            }
+            UpstairsAction::RepairCheck => {
+                self.on_repair_check().await;
+            }
+            UpstairsAction::Control(c) => {
+                self.on_control_req(c).await;
+            }
+        }
+
+        // Check to see whether live-repair can continue
+        //
+        // This must be called before acking jobs, because it looks in
+        // `Downstairs::ackable_jobs` to see which jobs are done.
+        if let Some(job_id) = self.downstairs.check_live_repair() {
+            let mut gw = self.guest.guest_work.lock().await;
+            self.downstairs.continue_live_repair(
+                job_id,
+                &mut gw,
+                &self.state,
+                self.generation,
+            );
+        }
+
+        // Send jobs downstairs as they become available.  This must be called
+        // after `continue_live_repair`, which may enqueue jobs.
+        for i in ClientId::iter() {
+            if self.downstairs.clients[i].should_do_more_work() {
+                self.downstairs.io_send(i).await;
+            }
+        }
+
+        // Handle any jobs that have become ready for acks
+        if self.downstairs.has_ackable_jobs() {
+            let mut gw = self.guest.guest_work.lock().await;
+            self.downstairs.ack_jobs(&mut gw, &self.stats).await;
+        }
+
+        // Check for client-side deactivation
+        if matches!(self.state, UpstairsState::Deactivating { .. }) {
```

Comment:
Putting it at the end made sense to me because it's after acking work + `retire_check`, so it's guaranteed to see the moment when a `Downstairs` empties out its queue (handled in `ack_jobs` right above this).

If we did deactivation checking before acking jobs, it would _probably_ still work (because it would see that all jobs were Done / Skipped, even if they weren't acked + removed from the list), but it's a little trickier to think about...